### PR TITLE
Add spread list and spread map operator support

### PR DESCRIPTION
**Pull request by sonallux**
This PR adds support for spread list `[*x, 1, 2]`, spread map `[*:x, a: 'a']` and spread in arguments `sum(*args, 42)`.

Closes [JENKINS-46163](https://issues.jenkins.io/browse/JENKINS-46163)

### CODE REVIEW

**Comment by sonallux** [comment]
@dwnusbaum and @car-roll friendly ping. This is ready for review.

**Comment by car-roll** [comment]
@sonallux thanks for the PR! This is probably more under @dwnusbaum 's expertise, but I'm going to take my time and learn what I can going through your PR.

**Review by dwnusbaum — CHANGES_REQUESTED** [review]
Thanks! I added a few comments. Any time we add support for new Groovy features I am concerned about the security and maintenance implications, but I think this is probably ok.

Can you file a linked PR to `workflow-cps` with a deployed snapshot of this PR?

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java [hunk 1]
```diff
@@ -60,7 +60,7 @@ public Next eval(Env e, Continuation k) {
 
         Object lhs;
         String name;
-        Object[] args = new Object[argExps.length];
+        List<Object> args = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (sonallux) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java]: Ok, I was not aware of that. I have changed to logic a bit to leave the type unchanged.

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadListBlockTest.groovy [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package com.cloudbees.groovy.cps.impl
+
+import com.cloudbees.groovy.cps.AbstractGroovyCpsTest
+import org.junit.Test
+
+class SpreadListBlockTest extends AbstractGroovyCpsTest {
```
[the remark this answers is not in the document]
Comment (sonallux) [reply lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadListBlockTest.groovy]: I have refactored all new tests to plain java test files.

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadListBlockTest.groovy [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Test
+    void simpleListSpreadEnd() {
+        Object result = evalCPS('''
+            def x = [3, 4, 5]
+            return [1, 2, *x]
+        ''')
+        assertEquals([1, 2, 3, 4, 5], result)
+    }
```
[the remark this answers is not in the document]
Comment (sonallux) [reply lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadListBlockTest.groovy]: Done 👍

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadMapBlockTest.groovy [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    void spreadEmptyMap() {
+        Object result = evalCPS("""
+            def y = [:]
+            return [*:y, a: 'hello']
+        """)
+        assertEquals([a: 'hello'], result)
+    }
+
+    @Test(expected = NullPointerException)
```
[the remark this answers is not in the document]
Comment (sonallux) [reply lib/src/test/groovy/com/cloudbees/groovy/cps/impl/SpreadMapBlockTest.groovy]: True, it's everywhere with explicit try/catch blocks. I have changed this.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/SpreadList.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    static SpreadList of(Object value) {
+        if (value == null) {
+            return new SpreadList(Collections.singleton(null));
+        }
+
+        if (value instanceof Collection) {
+            return new SpreadList((Collection<?>) value);
+        }
+        throw new SpreadListEvaluatingException("Cannot spread the list " + value.getClass().getName() + ", value " + value);
```
[the remark this answers is not in the document]
Comment (sonallux) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/SpreadList.java]: Thanks for finding this. I have changed the logic to match the behaviour of real groovy.

**Comment by sonallux** [comment]
> Thanks! I added a few comments. Any time we add support for new Groovy features I am concerned about the security and maintenance implications, but I think this is probably ok.

@dwnusbaum I have resolved all your comments.

> Can you file a linked PR to `workflow-cps` with a deployed snapshot of this PR?

Sure, did you mean the [workflow-cps-plugin](https://github.com/jenkinsci/workflow-cps-plugin)? I think I also need some advice on how to deploy `groovy-cps` to `repo.jenkins-ci.org`. Is it just `mvn deploy` to deploy a version?

**Comment by ajschmidt8** [comment]
Any updates on this PR?

**Comment by sonallux** [comment]
@dwnusbaum friendly ping. I have resolved all your comments.

**Comment by jycr** [comment]
Any news?

**Comment by jgsqware** [comment]
up pinging this too :D

**Comment by dwnusbaum** [comment]
I was able to look into this a bit this week. I filed https://github.com/cloudbees/groovy-cps/pull/123 with some improvements (mostly minor things). I am closing this PR in favor of that one.

**Pull request closed without merging**